### ISSUE TITLE
Deck crashed when server synchronization is impossible

### ISSUE THREAD

**Issue by neerro** [stefan-niedermann/nextcloud-deck#1063]
**Describe the bug**
See steps to reproduce field below

**Steps to reproduce the behavior:**
1. Within steps are described below, I had no capability to synchronized my Android Deck app to the server. I had access to internet on Android but haven't on server.
2. I've start Android deck app.
3. In the existing deck, I've tried to add a new card.
4. Then I saw error below. When I've tried to reproduce step 3 the error had repeated.

**Expected behavior**
A clear work app without bugs.

**Screenshots**
No

**Versions**
 - Nextcloud: 20.0.3
 - Nextcloud Deck: 1.4.2
 - Nextcloud Android: 3.16.1
 - Nextcloud Android Deck: 1.19.0

**Smartphone (please complete the following information):**
- Device: Honor 10 Light
- Android-Version: 10
- App-Store:
  - [x] Google Play Store

**Stacktrace**

```
App Version: 1.19.0
App Version Code: 1019000
Server App Version: 1.4.2
App Flavor: play

Files App Version Code: 30160190

---

OS Version: 4.14.116(10.0.0.221C10)
OS API Level: 29
Device: HWHRY-H
Manufacturer: HUAWEI
Model (and Product): HRY-LX1 (HRY-LX1RU)

---

com.nextcloud.android.sso.exceptions.UnknownErrorException: failed to connect to cloud.website.net/XX.XXX.XX.XXX (port 443) from /XX.XXX.XX.XXX (port 35533) after 60000ms: isConnected failed: ETIMEDOUT (Connection timed out)
	at com.nextcloud.android.sso.api.AidlNetworkRequest.performNetworkRequestV2(AidlNetworkRequest.java:172)
	at com.nextcloud.android.sso.api.NextcloudAPI.performNetworkRequestV2(NextcloudAPI.java:171)
	at com.nextcloud.android.sso.api.NextcloudAPI.lambda$performRequestObservableV2$1$NextcloudAPI(NextcloudAPI.java:110)
	at com.nextcloud.android.sso.api.-$$Lambda$NextcloudAPI$8WS9mapJbOYsPgbhBi25vBNYZqI.subscribe(Unknown Source:6)
	at io.reactivex.internal.operators.observable.ObservableFromPublisher.subscribeActual(ObservableFromPublisher.java:31)
	at io.reactivex.Observable.subscribe(Observable.java:12284)
	at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:32)
	at io.reactivex.Observable.subscribe(Observable.java:12284)
	at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:26)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.processTask(ThreadPoolExecutor.java:1187)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:929)

Full Crash:

App Version: 1.19.0
App Version Code: 1019000
App Flavor: play

Files App Version Code: 30160190

---

OS Version: 4.14.116(10.0.0.221C10)
OS API Level: 29
Device: HWHRY-H
Manufacturer: HUAWEI
Model (and Product): HRY-LX1 (HRY-LX1RU)

---

java.lang.IllegalStateException: Fragment NewCardDialog{84ae0cf} (4fd6e8ad-cdbc-4f7f-8815-b8a39c50587f) has not been attached yet.
	at androidx.fragment.app.Fragment.getChildFragmentManager(Fragment.java:1075)
	at it.niedermann.nextcloud.deck.ui.branding.NewCardDialog$2.lambda$onError$1$NewCardDialog$2(NewCardDialog.java:179)
	at it.niedermann.nextcloud.deck.ui.branding.-$$Lambda$NewCardDialog$2$6wrtL8e1qYlISwDIbL7-7X8UMs8.run(Unknown Source:4)
	at android.os.Handler.handleCallback(Handler.java:888)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loop(Looper.java:213)
	at android.app.ActivityThread.main(ActivityThread.java:8178)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1101)
```

**Comment by desperateCoder** [comment]
Hi @neerro, thanks for your report and sorry for the inconveniences! This is quite exotic and we didn't test this specific case yet, but that's a great hint, thanks for that!

We'll have a look at it, I guess we'll catch that and bring up a readable error, but the app should work as expected, you're totally right.

We'll keep you up to date!

**Comment by neerro** [comment]
It's impossible to work with Android Deck  app when server is not response for synchronization as I've got the next error when I have tryied create a new card in existing deck.
After entering header name and waiting for a some time (not more 15 sec.) I met the next error and crashing app. 

```
App Version: 1.19.0
App Version Code: 1019000
Server App Version: 1.4.3
App Flavor: play

Files App Version Code: 30160190

---

OS Version: 4.14.116(10.0.0.221C10)
OS API Level: 29
Device: HWHRY-H
Manufacturer: HUAWEI
Model (and Product): HRY-LX1 (HRY-LX1RU)

---

com.nextcloud.android.sso.exceptions.UnknownErrorException: failed to connect to cloud.khilov.net/79.120.11.229 (port 443) from /100.90.36.40 (port 59867) after 60000ms: isConnected failed: ETIMEDOUT (Connection timed out)
	at com.nextcloud.android.sso.api.AidlNetworkRequest.performNetworkRequestV2(AidlNetworkRequest.java:172)
	at com.nextcloud.android.sso.api.NextcloudAPI.performNetworkRequestV2(NextcloudAPI.java:171)
	at com.nextcloud.android.sso.api.NextcloudAPI.lambda$performRequestObservableV2$1$NextcloudAPI(NextcloudAPI.java:110)
	at com.nextcloud.android.sso.api.-$$Lambda$NextcloudAPI$8WS9mapJbOYsPgbhBi25vBNYZqI.subscribe(Unknown Source:6)
	at io.reactivex.internal.operators.observable.ObservableFromPublisher.subscribeActual(ObservableFromPublisher.java:31)
	at io.reactivex.Observable.subscribe(Observable.java:12284)
	at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:32)
	at io.reactivex.Observable.subscribe(Observable.java:12284)
	at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:26)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.processTask(ThreadPoolExecutor.java:1187)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:929)

```

**Comment by neerro** [comment]
The same error I've got when I tryied renew synchronization when server access is out.

Step to reproduce:

I open existing deck, put finger on mobile device screen and pull down to renew. After that round spinner was appeared but without any animation, it seems spinner had freezed. Then I've got message below  :(

```
App Version: 1.19.0
App Version Code: 1019000
Server App Version: 1.4.3
App Flavor: play

Files App Version Code: 30160190

---

OS Version: 4.14.116(10.0.0.221C10)
OS API Level: 29
Device: HWHRY-H
Manufacturer: HUAWEI
Model (and Product): HRY-LX1 (HRY-LX1RU)

---

com.nextcloud.android.sso.exceptions.UnknownErrorException: failed to connect to cloud.website.net/XX.XXX.XX.XXX (port 443) from /XXX.XX.XX.XX (port 56465) after 60000ms: isConnected failed: EHOSTUNREACH (No route to host)
	at com.nextcloud.android.sso.api.AidlNetworkRequest.performNetworkRequestV2(AidlNetworkRequest.java:172)
	at com.nextcloud.android.sso.api.NextcloudAPI.performNetworkRequestV2(NextcloudAPI.java:171)
	at com.nextcloud.android.sso.api.NextcloudAPI.lambda$performRequestObservableV2$1$NextcloudAPI(NextcloudAPI.java:110)
	at com.nextcloud.android.sso.api.-$$Lambda$NextcloudAPI$8WS9mapJbOYsPgbhBi25vBNYZqI.subscribe(Unknown Source:6)
	at io.reactivex.internal.operators.observable.ObservableFromPublisher.subscribeActual(ObservableFromPublisher.java:31)
	at io.reactivex.Observable.subscribe(Observable.java:12284)
	at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:26)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.processTask(ThreadPoolExecutor.java:1187)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:929)

```

**Comment by stefan-niedermann** [comment]
So, let me summarize, what we got until now:

1. The app did not crash, it displayed a message, that the synchronization was not successful
2. Your Nextcloud instance is not reachable

So what is your concrete suggestion about how to handle this state?

If a Nextcloud instance is not reachable, then the user gets a notification - that looks correct from my point of view. Do you suggest to hide this information from the user? The problem is, that tze next ticket will be "Deck does not sync and i have no idea why" obviously, so this is not a good solution in my opinion.

If the synchronization is not successful, we should give feedback to the user, no?

Keep in mind, that "Offline capabilities" means that your *device* is offline - not your instance, your internet provider or anything else.

In which way is it a permanent state for you, that your server is offline for a longer period of time?

**Comment by neerro** [comment]
> 
> 
> So, let me summarize, what we got until now:
> 
>     1. The app did not crash, it displayed a message, that the synchronization was not successful
>
Not exactly, as sometimes I can see collapse the app after I push on close button in error window.
> 
>     2. Your Nextcloud instance is not reachable
>
Within that issue, yes.
> 
> 
> So what is your concrete suggestion about how to handle this state?
> 
> If a Nextcloud instance is not reachable, then the user gets a notification - that looks correct from my point of view. Do you suggest to hide this information from the user? The problem is, that tze next ticket will be "Deck does not sync and i have no idea why" obviously, so this is not a good solution in my opinion.
>
My suggestion in that case is very simple. Information about synchronization should be processed in silent mode. For example, as it is represented in this time. Round arrow spinner in each card are shown the status. For example red color non synchronized, green color - everything is ok. If not synchronized, red spinner will be visible until synchronization will be done, green spinner will be disappear after some time.
> 
> If the synchronization is not successful, we should give feedback to the user, no?
>
Yes, definitely. If it is an internal issue with app we can show it in pop-up window, or we can add a log menu to be possible for export. If app could not be synchronized with server during the server is off or inaccessible, just wait for time when server will be online or internet connection on mobile device will appear. The same steps has implemented in similar apps.
>
> Keep in mind, that "Offline capabilities" means that your _device_ is offline - not your instance, your internet provider or anything else.
>
The situation in our life could be different, and we need to prepare the app for varied scenarios.
>
> In which way is it a permanent state for you, that your server is offline for a longer period of time?
>
Yes, some period server has offline mode. And I wanna work with android deck also offline and then when server will be online I would like synchronized my work which were done in offline mode in android deck app.
>

**Comment by stefan-niedermann** [comment]
All right, based on your feedback, we decided to catch exceptions which are most likely related to temporary infrastructure, connectivity problems or a unreachable server in the future silently. The sync indicators will stay of course to signal the user that an item is not yet synchronized.

When performing an active pull to refresh (swiping down), we will continue to display those kinds of errors, because we believe that the user *needs* an active information in case he actively wants to synchronize.

This new behavior should also fix the scenarios you described in #1064 as well as #1061 and will be shipped with Deck `1.19.1`

**Comment by stefan-niedermann** [comment]
Deck Android 1.19.2 has been rolled out today for all users, closing this issue therefore 🙂 

Thank you for the productive discussion.
